osquery/dispatcher/posix/query_profiler.cpp

```diff
@@ -0,0 +1,162 @@
+#define _GNU_SOURCE
+#endif
+#endif
+
+#include <cerrno>
+#include <cstdint>
+#include <cstring>
+
+#include <sys/resource.h>
+#include <sys/time.h>
+
+#include <boost/format.hpp>
+#include <boost/io/detail/quoted_manip.hpp>
+
+#include <osquery/dispatcher/query_profiler.h>
+#include <osquery/killswitch.h>
+#include <osquery/logger.h>
+#include <osquery/numeric_monitoring.h>
+
+namespace osquery {
+namespace {
+
+int getRusageWho() {
+  return
+#ifdef __linux__
+      RUSAGE_THREAD; // Linux supports more granular profiling
+#else
+      RUSAGE_SELF;
+#endif
+}
+
+void recordRusageStatDifference(int64_t start_stat,
+                                int64_t end_stat,
+                                const std::string& stat_name) {
+  if (end_stat == 0) {
+    TLOG << "rusage field " << boost::io::quoted(stat_name)
+         << " is not supported";
+  } else if (start_stat <= end_stat) {
+    monitoring::record(
+        stat_name, end_stat - start_stat, monitoring::PreAggregationType::P50);
+  } else {
+    LOG(WARNING) << "Possible overflow detected in rusage field: "
+                 << boost::io::quoted(stat_name);
+  }
+}
+
+void recordRusageStatDifference(const struct timeval& start_stat,
+                                const struct timeval& end_stat,
+                                const std::string& stat_name) {
+  recordRusageStatDifference(
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::seconds(start_stat.tv_sec) +
+          std::chrono::microseconds(start_stat.tv_usec))
+          .count(),
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::seconds(end_stat.tv_sec) +
+          std::chrono::microseconds(end_stat.tv_usec))
+          .count(),
+      stat_name + ".millis");
+}
+
+void recordRusageStatDifference(const struct rusage& start_stats,
+                                const struct rusage& end_stats,
+                                const std::string& monitoring_path_prefix) {
+  recordRusageStatDifference(
+      0, end_stats.ru_maxrss, monitoring_path_prefix + ".rss.max.kb");
+
+  recordRusageStatDifference(start_stats.ru_maxrss,
+                             end_stats.ru_maxrss,
+                             monitoring_path_prefix + ".rss.increase.kb");
+
+  recordRusageStatDifference(start_stats.ru_inblock,
+                             end_stats.ru_inblock,
+                             monitoring_path_prefix + ".input.load");
+
+  recordRusageStatDifference(start_stats.ru_oublock,
+                             end_stats.ru_oublock,
+                             monitoring_path_prefix + ".output.load");
+
+  recordRusageStatDifference(start_stats.ru_utime,
+                             end_stats.ru_utime,
+                             monitoring_path_prefix + ".time.user");
+
+  recordRusageStatDifference(start_stats.ru_stime,
+                             end_stats.ru_stime,
+                             monitoring_path_prefix + ".time.system");
```

Comment:
Do we care about system time here? IMHO, the most important time parameter of running query is a user time and full time. If this two is fine we likely don't care about system time. Otherwise we can repeat the intended query in controllable environment and figure out more parameters. Does it makes sense for you?

Reply:
I see no need of trade offs here. As an example, system time gives information about how heavy syscalls we are using.

Reply:
Oh, I'm just talking about more keys to send. For some monitoring systems it could be a problem :)

Reply:
What system are we talking about specifically? I don't think we should lose information, without very specific/practical limitation.